### PR TITLE
Extension and mail agent (WIP)

### PR DESCRIPTION
**Pull request by fureev**

### CODE REVIEW

**Comment by fureev** [comment]
@alexeyshockov Создал для обсуждения, еще не дописано и не работает

**Comment by alexeyshockov** [comment]
For what purpose you change file permissions (composer.lock & composer.json)?

**Comment by alexeyshockov** [comment]
Great work on integration with Behat! But I suggest some additional model change.

Extract accounts to separate class (Account). After this refactoring we will have 3 “content” classes — Mailbox, Message and Account, and one “functional” class — MailAgent. Mail agent will contains all ezComponents object inside, and other classes will be abstracted from this.

**Comment by fureev** [comment]
Because i can't edit files in vendor folder.

**Comment by fureev** [comment]
It works!

**Comment by alexeyshockov** [comment]
Excellent! So many refactoring in one PR :)

**Comment by fureev** [comment]
:sunglasses:

File: src/Staffim/Behat/MailExtension/Account.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * @var string
+     */
+    private $password;
+
+    /**
+     * @param $server string server name
+     * @param $auth array of ['login': login, 'password': password]
+     */
+    public function __construct($server, $auth=['login' => '', 'password' => ''])
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/Account.php]: Please make login and password as parameters

File: src/Staffim/Behat/MailExtension/Account.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        $this->server = $server;
+        $this->user = $auth['login'];
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/Account.php]: I'm think field `login` is preferable

File: src/Staffim/Behat/MailExtension/Context/MailAwareInterface.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+     * @return mixed
+     */
+    public function setMailAgent(MailAgent $mailAgent);
+
+    /**
+     * Sets parameters provided for MailAgent.
+     *
+     * @param array $parameters
+     */
+    public function setMailAgentParameters(array $parameters);
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/Context/MailAwareInterface.php:23]: Is parameters are really needed here?

File: src/Staffim/Behat/MailExtension/Context/RawMailContext.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    public function iSignInToSmtpServer($mailServer, $user, $password)
+    {
+        $smtpAccount = new Account($mailServer, ['login' => $user, 'password' => $password]);
+        $this->getMailAgent()->connectSmtpServer($smtpAccount);
+    }
+
+    /**
+     * @When /^(?:|I )sign out from mail server$/
+//     * @When /^(?:|я )выхожу с почтового сервера$/
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/Context/RawMailContext.php]: Remove comments

File: src/Staffim/Behat/MailExtension/MailAgent.php [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    /**
+     * @param Account $smtpAccount
+     */
+    public function setSmtpAccount($smtpAccount)
+    {
+        $this->smtpAccount = $smtpAccount;
+    }
+
+    /**
+* @param AccountInterface $pop3Account
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/MailAgent.php]: Please fix identation

File: src/Staffim/Behat/MailExtension/MailAgent.php [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+     */
+    public function __construct(AccountInterface $pop3Account, AccountInterface $smtpAccount = null)
+    {
+        $this->pop3Account = $pop3Account;
+
+        $this->pop3Transport = new \ezcMailPop3Transport($pop3Account->getServerName());
+
+        $this->mailParser = new \ezcMailParser();
+
+        if ($smtpAccount) {
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/MailAgent.php]: That condition is no needed

File: src/Staffim/Behat/MailExtension/MailAgent.php [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    public function createReplyMessage($mail, $text)
+    {
+        $replyMail = \ezcMailTools::replyToMail($mail, $mail->to[0]);
+        $replyMail->body = new \ezcMailText($text, 'utf8', '8bit', 'utf8');
+
+        return $replyMail;
+    }
+
+//    TODO
+    public function addAttach($mail, $filepath)
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/MailAgent.php]: Add attach into mail agent?

File: src/Staffim/Behat/MailExtension/MailAgent.php [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+     * Receive messages to mailbox
+     */
+    public function receive()
+    {
+        $this->connectPop3Server();
+
+        $mails = $this->pop3Transport->fetchAll();
+        $mails = $this->mailParser->parseMail($mails);
+
+        $this->setMailbox(new Mailbox($mails));
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/MailAgent.php]: Maybe `receive()` should return `mailbox`?

File: src/Staffim/Behat/MailExtension/Message.php [hunk 9]
```diff
@@ -157,7 +126,7 @@ public function serializeAddressHeaders()
      */
     public function getPlainMessage()
     {
-        return explode('Content-Type: text/html', $this->mail->generate())[0];
+        return "Plain message:\n" . explode('Content-Type: text/html', $this->mail->generate())[0];
```
Comment (megazoll) [inline src/Staffim/Behat/MailExtension/Message.php]: Oops?

File: src/Staffim/Behat/MailExtension/Extension.php [hunk 10]
```diff
@@ -5,9 +5,11 @@
 use Behat\Behat\Extension\ExtensionInterface;
 
 use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
+use Symfony\Component\Config\FileLocator;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
 
-class Extension implements ExtensionInterface
+class Extension extends \Behat\Behat\Extension\Extension implements ExtensionInterface
```
Comment (alexeyshockov) [inline src/Staffim/Behat/MailExtension/Extension.php]: `use Behat\Behat\Extension\Extension`

File: src/Staffim/Behat/MailExtension/Extension.php [hunk 11]
```diff
@@ -15,7 +17,19 @@ class Extension implements ExtensionInterface
      */
     public function load(array $config, ContainerBuilder $container)
     {
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
+        $loader->load('ExtensionServices.yml');
 
+        foreach ($config as $ns => $tlValue) {
```
Comment (alexeyshockov) [inline src/Staffim/Behat/MailExtension/Extension.php]: Please, expand `$ns` and `$tlValue` to more readable names. At this I can not imagine, what they might mean.

File: src/Staffim/Behat/MailExtension/Message.php (hunk 9 above)
Comment (alexeyshockov) [inline src/Staffim/Behat/MailExtension/Message.php]: And small note: please, don't use array dereferencing (we want support PHP 5.3, are we?).

File: src/Staffim/Behat/MailExtension/Extension.php (hunk 11 above)
Comment (fureev) [inline src/Staffim/Behat/MailExtension/Extension.php]: too long value

File: src/Staffim/Behat/MailExtension/Context/MailAwareInterface.php (hunk 3 above)
Comment (fureev) [inline src/Staffim/Behat/MailExtension/Context/MailAwareInterface.php:23]: There aren't any other mail extension parameters in context

File: src/Staffim/Behat/MailExtension/MailAgent.php (hunk 7 above)
Comment (fureev) [inline src/Staffim/Behat/MailExtension/MailAgent.php]: Add attach to mail message in mail agent

**Comment by fureev** [comment]
TODO add wait func in mail Agent Interface

**Pull request merged**